Prevent visiting the same address too soon after the previous visit

begedin opened the pull request:
- [Asana task: Same door can't be knocked on within X time span](https://app.asana.com/0/60127409159606/60127409159733)
- Also resolves [asana task Default visit factory creates a visit with 2 people updated. Should probably move that to a trait.](https://app.asana.com/0/60127409159606/61546570380056)
# Note

@JoshSmith I'm aware of your comment in the asana task:

> For env I meant Rails.env, not an ENV variable.

However, I wasn't sure what you meant by that. 

In any case, I'm hoping that part of the changeset will be minor, so I implemented this with the assumption we'll be looking for the interval configuration setting from within ENV. This is why the `.env.sample` file has that added line.
# Implementation
- ~~Added a `has_many :address_updates` relationship to the `Address` model (includes spec in the model spec module).~~
- ~~Added a `created_at` timestamp to the `:address_updates` table. (includes spec in the model spec module).~~
- Added a `recently_visited?` instance method to the `Address` model. This fetches the configurable interval from the ENV and determines if the ~~any of the addresses `address_updates` were created~~ if `address.visited_at` is in that interval.
- Defined a custom error `VisitNotAllowed` with a fixed message `"You cannot visit this address so soon since it was last visited"`.
- The `Address.new_or_existing_from_params` class method calls `Address.existing_with_params` private class method. This method raises the new custom exception after initializing the address, by checking `address.recently_visited?`. I think the clearest place to put this assertion would actually be at the top level, within the visit scenario, but putting it here will avoid the most work if the exception ends up being raised. **I could use a second opinion here, though. Should I move the assertion and exception raising to a higher level, into the `CreateVisit` scenario.
- The `ErrorSerializer` now also supports the new `VisitNotAllowed` error
- The `CreateVisit` scenario now also catches the new `VisitNotAllowed` error
## Other additions and changes
- There's a new transient attribute in the default address factory, called `recently_visited`. By default, it's false. If it's set to true and there is no explicitly defined `visited_at` value, it will set it to a value within the invalid datetime range.
- I had to move the default `people_updates` functionality of the visit factory into a `:with_people` trait. Otherwise, there would have been too much clashing and additionally, the trait does make some things clearer.

begedin committed:
Feature: Throw an error when visiting the same address within a configurable time period

begedin committed:
Updated address model specs

begedin committed:
Filled in missing specs

begedin committed:
Cleanup in user spec (duplicate tests)

begedin committed:
Cleanup address spec a bit

begedin commented:
I'm expecting changes in this one, but I'm at a point where I need a review to continue.

joshsmith commented on lib/ground_game/errors/visit_not_allowed.rb:
```diff
@@ -0,0 +1,9 @@
+      super ERROR_MESSAGE
+    end
+    private
+      ERROR_MESSAGE = "You cannot visit this address so soon since it was last visited"
```
I'll change this to something else. Maybe "You can't submit the same address so quickly after it was last visited."

joshsmith commented on app/models/address.rb:
```diff
@@ -22,12 +24,20 @@ class Address < ActiveRecord::Base
+    upper_bound = (DateTime.now.to_i)
+    invalid_interval = lower_bound..upper_bound
+
+    self.address_updates.any? { |update| invalid_interval.include? update.created_at.to_i }
```
Why not use the last address update an address had? Seems like we're querying for too much here.

joshsmith replied:
Even better, why not add a cached column for the last time an address received an update? Then once you have the address you're not doing any more queries. This puts all the work on the write and very little on the read/processing side of things.

begedin replied:
Looking at the code, we already have such a column - `visited_at`. We had it defined a while ago, but never really used it. There are only specs that ensure it exists, nothing more. 

I could do two tings here
1. Rename it to `update_at` and make use of it being set automatically by the timestamp functionality. However, that would mean the column would also update in cases where we somehow update the address directly, not through creating a visit. We can't really do that at the moment, but still, we may be doing it at some point in the future.
2. Leave it named `visited_at` and set it manually when an address update for that address is created. 

Regardless of which option we choose, there's the matter of the timestamp field I added to `AdddressUpdate` in this PR. Should I revert that change or keep it? We don't need it at the moment, but we may in the future.

joshsmith commented:
Let's make some minor adjustments here, and after that, good to go. :+1:

begedin committed:
Merge branch 'development' into feature/limit-multiple-visits-timespan

Conflicts:
	app/models/address.rb
	spec/lib/ground_game/scenario/create_visit_spec.rb

begedin committed:
Switched to using address.visited_at to determine if address was recently visited

begedin committed:
Changed error message for VisitNotAllowed

begedin commented:
@JoshSmith I made the changes
- Added a custom error we raise in the case of `InvalidBestCanvasResponse`. It makes the code slightly more readable, IMO, and it doesn't add a lot of overhead. It was not as necessary when it was just that one error, but with this pr, we also raise `VisitNotAllowed` within the same code block.
- Changed the error message we use to the one you asked for.
- Switched from checking if there is an `address_update` with `created_at` within the time range, to simply checking `address.visited_at`. This field was present for a long time, but was never being updated, so I changed the `CreateVisit` scenario to update it and added specs to make sure this is happening.

begedin committed:
Merge branch 'development' into feature/limit-multiple-visits-timespan

Conflicts:
	.env.sample
	app/serializers/error_serializer.rb
	lib/ground_game/scenario/create_visit.rb
	spec/serializers/error_serializer_spec.rb

joshsmith committed:
Merge

joshsmith committed:
Add in min interval access

joshsmith committed:
Add notes to README

joshsmith committed:
Merge

joshsmith committed:
Merge in development AGAIN

joshsmith committed:
Update visit factory

joshsmith committed:
Add association for person update factory

estiens committed:
refactorings to fix tests and factories

joshsmith committed:
Add to_i
